#102 - Validate query from frontend against json schema

michael-82 opened the pull request:
**What's in the PR**
* the [json schema for structured query v2](https://github.com/num-codex/codex-structured-query/blob/structured-query-v2/structured-query/json-schema/2021_10_18_StructuredQueryV2Schema.json) was added and is used to validate incoming queries on the /run-query endpoint against it
* if validation fails, a 422 error is thrown - otherwise nothing changes
* **breaking change**: the run-query method was returning strange responses, always with status code 200 ok and the real code somewhere in the response json body. this was changed to a regular response. However, this currently breaks compatibility with the frontend. @thkoehler11 was involved in the decision to still change it and then adapt the frontend to it.

**How to test manually**
* there is a test class that checks if several schema violations are correctly detected
* otherwise, use a rest client of your choice and check if valid and invalid queries are correctly handled.

**Examples**
Please note that the url may be different on your setup

Call with valid query:
```sh
curl --request POST \
  --url http://localhost:8090/api/v1/query-handler/run-query \
  --header 'Content-Type: application/json' \
  --data '{
  "version": "http://to_be_decided.com/draft-2/schema#",
  "display": "Beispiel-Query",
  "inclusionCriteria": [
    [
      {
        "termCodes": [
          {
            "code": "LL2191-6",
            "display": "Geschlecht",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "concept",
          "selectedConcepts": [
            {
              "code": "F",
              "display": "female",
              "system": "https://fhir.loinc.org/CodeSystem/$lookup?system=http://loinc.org&code=LL2191-6",
              "version": ""
            },
            {
              "code": "M",
              "display": "male",
              "system": "https://fhir.loinc.org/CodeSystem/$lookup?system=http://loinc.org&code=LL2191-6",
              "version": ""
            }
          ]
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "30525-0",
            "display": "Alter",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-comparator",
          "comparator": "gt",
          "unit": {
            "code": "a",
            "display": "Jahr"
          },
          "value": 18
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "F00",
            "display": "F00",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ]
      },
      {
        "termCodes": [
          {
            "code": "F09",
            "display": "F09",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ],
        "timeRestriction": {
          "beforeDate": "2021-10-09",
          "afterDate": "2021-09-09"
        }
      }
    ]
  ],
  "exclusionCriteria": [
    [
      {
        "termCodes": [
          {
            "code": "LL2191-6",
            "display": "Geschlecht",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "concept",
          "selectedConcepts": [
            {
              "code": "male",
              "display": "male",
              "system": "",
              "version": ""
            }
          ]
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "30525-0",
            "display": "Alter",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-comparator",
          "comparator": "gt",
          "unit": {
            "code": "year",
            "display": "Jahr"
          },
          "value": 65
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "F00.9",
            "display": "F00.9",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ]
      },
      {
        "termCodes": [
          {
            "code": "8310-5",
            "display": "Körpertemperatur",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-range",
          "unit": {
            "code": "Cel",
            "display": "°C"
          },
          "minValue": 35,
          "maxValue": 39
        },
        "attributeFilters": [
          {
            "attributeCode": {
              "code": "method",
              "system": "abide",
              "display": "method"
            },
            "type": "concept",
            "selectedConcepts": [
              {
                "code": "LA9370-3",
                "system": "http://loinc.org",
                "display": "Axillary"
              }
            ]
          }
        ],
        "timeRestriction": {
          "beforeDate": "2021-10-09",
          "afterDate": "2021-09-09"
        }
      }
    ]
  ]
}
'
```
Call with invalid query (no inclusion criteria present):
```sh
curl --request POST \
  --url http://localhost:8090/api/v1/query-handler/run-query \
  --header 'Content-Type: application/json' \
  --data '{
  "version": "http://to_be_decided.com/draft-2/schema#",
  "display": "Beispiel-Query",
  "exclusionCriteria": [
    [
      {
        "termCodes": [
          {
            "code": "LL2191-6",
            "display": "Geschlecht",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "concept",
          "selectedConcepts": [
            {
              "code": "male",
              "display": "male",
              "system": "",
              "version": ""
            }
          ]
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "30525-0",
            "display": "Alter",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-comparator",
          "comparator": "gt",
          "unit": {
            "code": "year",
            "display": "Jahr"
          },
          "value": 65
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "F00.9",
            "display": "F00.9",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ]
      },
      {
        "termCodes": [
          {
            "code": "8310-5",
            "display": "Körpertemperatur",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-range",
          "unit": {
            "code": "Cel",
            "display": "°C"
          },
          "minValue": 35,
          "maxValue": 39
        },
        "attributeFilters": [
          {
            "attributeCode": {
              "code": "method",
              "system": "abide",
              "display": "method"
            },
            "type": "concept",
            "selectedConcepts": [
              {
                "code": "LA9370-3",
                "system": "http://loinc.org",
                "display": "Axillary"
              }
            ]
          }
        ],
        "timeRestriction": {
          "beforeDate": "2021-10-09",
          "afterDate": "2021-09-09"
        }
      }
    ]
  ]
}
'
```

michael-82 committed:
#102 - Validate query from frontend against json schema

- add first version of json validation with pre packaged schema file
- change http responses in run-query endpoint since it always returned 200

michael-82 committed:
#102 - Validate query from frontend against json schema

- add javadoc

michael-82 committed:
#102 - Validate query from frontend against json schema

- update json schema dependency to latest version (different artifact and group id)
- set json schema draft version to 7 to support date format validation
- add QueryValidationTest

DiCanio reviewed (changes requested):
Maybe we can even shift to a more "Spring" approach by using the following dependency:
https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-validation

With this most of the implementation logic can be kept. Making use of it is simply a matter of adding a `@Valid` annotation to the appropriate endpoint field and implementing a constraint validator (simply re-use the implementation logic used in `QueryValidator`). Doing so even takes care of translating any error to the correct status codes (and response in general).

michael-82 committed:
Update docker-compose.yml

Rename env variable and set default to true

Co-authored-by: DiCanio <DiCanio@users.noreply.github.com>

michael-82 committed:
#102 - Validate query from frontend against json schema

- move schema creation to QueryValidatorSpringConfig
- use @Valid annotation and implement custom validator
- add spring-boot-starter-validation dependency

michael-82 committed:
Merge branch 'develop' into feature/102-Validate-query-from-frontend-against-json-schema

michael-82 commented on src/main/java/de/numcodex/feasibility_gui_backend/query/QueryHandlerRestController.java:
```diff
@@ -26,34 +30,56 @@
+  public ResponseEntity<Object> runQuery(
       @RequestBody StructuredQuery query, @Context HttpServletRequest httpServletRequest) {
 
+    if (queryValidationEnabled) {
```
Not really sure if I exactly got what you meant. However I moved the "enabled" check to the validator itself

michael-82 commented on src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidator.java:
```diff
@@ -0,0 +1,50 @@
+          .schemaJson(jsonSchema)
+          .draftV7Support()
+          .build();
+      var schema = loader.load().build();
```
moved it there

michael-82 commented on src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidatorSpringConfig.java:
```diff
@@ -0,0 +1,14 @@
+public class QueryValidatorSpringConfig {
+
+    @Bean
+    public QueryValidator createQueryValidator(ObjectMapper objectMapper) {
```
I changed it to create a new one

michael-82 commented on src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidator.java:
```diff
@@ -0,0 +1,50 @@
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+@Slf4j
```
my bad - it is used now

michael-82 commented:
> Maybe we can even shift to a more "Spring" approach by using the following dependency: https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-validation
> 
> With this most of the implementation logic can be kept. Making use of it is simply a matter of adding a `@Valid` annotation to the appropriate endpoint field and implementing a constraint validator (simply re-use the implementation logic used in `QueryValidator`). Doing so even takes care of translating any error to the correct status codes (and response in general).

I changed it accordingly. Not 100% sure if this is the correct/best way to do it though.

DiCanio committed:
Move and Rename to Clarify Association with API

This is simply done in order to group what needs
to go together. Additionally, renames several
classes so that they have a StructuredQuery prefix.
Without this adjustment we are running into a
mental naming clash since there is also the concept
of an internal query. Having the prefix clarifies
the context.

DiCanio committed:
Switch to Startup Configuration for SQ Validation

Removes any runtime checks associated with structured
query validation. Checks are moved so that they are
evaluated during application startup.

DiCanio committed:
Add Validation Test for Query Rest Controller

Add tests for enabled structured query validation
on a REST controller level.

DiCanio committed:
Revise Test

Use lambda expressions where applicable and use a
JUnit 5 compatible Mockito extension.

DiCanio commented:
I made some changes. Feel free to check whether you agree with them.

- rename and move classes to the API package (clarifies use case)
- switch to startup configuration (no runtime checks within the validator)
- add tests on a REST controller level (they are important)
- revise some exception handling passages

Furthermore, we do not have a test case on a REST controller level for when validation is turned off. I'm still not happy with that unfounded decision of being able to turn validation off. At this point I repeatedly asked for the reasoning behind this without ever getting a technical answer. Thus, please simply go ahead and take care of any potential fallout.

DiCanio commented:
Just noticed, that we also need an entry within the README documenting the validation switch.

Other than that, please see the comment above.

michael-82 committed:
#102 - Validate query from frontend against json schema

- add description of QUERY_VALIDATION_ENABLED parameter to README.md
